kore/src/Kore/Builtin/Set.hs

```diff
@@ -279,12 +283,35 @@ evalIn =
                     case arguments of
                         [_elem, _set] -> (_elem, _set)
                         _ -> Builtin.wrongArity Set.inKey
-            _elem <- hoistMaybe $ Builtin.toKey _elem
-            _set <- expectConcreteBuiltinSet Set.inKey _set
-            (Builtin.appliedFunction . asExpandedBoolPattern)
-                (Map.member _elem _set)
+            let setSymbolic = do
+                    _elem <- hoistMaybe $ Builtin.toKey _elem
+                    _set' <- expectBuiltinSet Set.inKey _set
+                    let result = Domain.isConcreteKeyOfAc _elem _set'
+                    ifTrueAddDefinednessCondition result _set
+                bothSymbolic = do
+                    _set' <- expectBuiltinSet Set.inKey _set
+                    let result = Domain.isSymbolicKeyOfAc _elem _set'
+                    ifTrueAddDefinednessCondition result _set
+                bothConcrete = do
+                    _elem <- hoistMaybe $ Builtin.toKey _elem
+                    _set <- expectConcreteBuiltinSet Set.inKey _set
+                    (Builtin.appliedFunction . asExpandedBoolPattern)
+                        (Map.member _elem _set)
+            setSymbolic <|> bothSymbolic <|> bothConcrete
       where
         asExpandedBoolPattern = Bool.asPattern resultSort
+        ifTrueAddDefinednessCondition result setTerm =
```

Comment:
Could we name this function something else, perhaps `returnIfTrueAndDefined`? To me, the name sounds like this function always returns a result, but only adds the definedness condition if the result is `True`, which would be wrong for several reasons. I had to read it several times before I understood. :laughing: